compiler/resolution/implementForallIntents2.cpp

```diff
@@ -0,0 +1,1097 @@
+      startup, combine() then delete it at task teardown
+      (do not do this "at the top level")
+
+    - for a RI, set up + tear down the shadow variable to yield
+      based on the current reduce op
+
+    - for a TPV (task-private variable), set up + tear down the shadow
+      variable to yield (no corresponding actual+formal in above steps)
+
+IF THE YIELD IS INSIDE A FORALL
+
+  * augment the ForallStmt with forall intents
+
+  * for a RI, skip certain steps
+
+IF THE YIELD IS INSIDE A FOR OVER A PARALLEL ITERATOR
+
+  * convert to forall loop
+    (preserving the semantics)
+
+  * process the forall as above
+
+*/
+
+/////////// helper data structures ///////////
+
+//
+// FIinfo aka Forall Intent info
+//
+// Contains all things pertinent to a single forall intent
+// across the call described by the enclosing FIcontext.
+//
+// RI = Reduce Intent  TPV = Task-Private Variable (upcoming)
+//
+class FIinfo {
+public:
+  // The shadow var in the ForallStmt being processed.
+  // In most cases it can be used only for info ex. the intent.
+  ShadowVarSymbol* svSymbol;
+
+  // The variable to be passed as an actual
+  // to the parallel iterator, task function,
+  // or ForallStmt within the parallel iterator.
+  // NULL for a TPV (todo soon: always NULL in this case?)
+  Symbol* fiActual;
+
+  // The corresponding formal or (for a TPV) local variable.
+  // For a RI, this is "parentOp"; otherwise the shadow variable.
+  // For a TPV, this is a local variable, once created;
+  // otherwise the formal of the parallel iterator or task fn.
+  // Todo: better name for it?
+  Symbol* fiFormal;
+
+  // For a reduce intent - non-NULL once created:
+  Symbol* riCurrentOp;
+  Symbol* riShadowVar;
+
+  // constructor
+  // invoked implicitly upon FIcontext::fiInfos.resize()
+  FIinfo() : svSymbol(0), fiActual(0), fiFormal(0),
+             riCurrentOp(0), riShadowVar(0) { }
+};
+
+typedef std::vector<FIinfo> FIvec;
+
+//
+// FIcontext aka Forall Intents contest
+//
+// Contains things pertinent to the 'call' being traversed
+// during extendLeaderNew(). The 'call' can be either:
+//
+//   when inIterator(): to the parallel iterator, from the ForallStmt,
+//
+//   when inTaskFn(): to a task function,
+//     from within the parallel iterator or its task function.
+//
+// Multiple yield statements may be handled/extended
+// under the same FIcontext.
+//
+class FIcontext {
+public:
+  // The forall stmt being processed.
+  ForallStmt*  forall;
+
+  // The call in being looked at.
+  CallExpr*    call;
+
+  // The iterator or task function being called.
+  FnSymbol*    curFn;
+
+  // Calling a function nested in the iterator?
+  bool         nested;
+
+  // The iterator's return symbol.
+  VarSymbol*   retSym;
+
+  // The FnSymbol for task startup/teardown code, usually 'curFn'.
+  // It is an enclosing function when descending into a task function
+  // that does not create a user-level task. Such a task function
+  // does not affect the intents.
```

Comment:
I'm having trouble understanding this comment. wouldn't you want to add task startup/teardown code within the body of the task?